library/src/main/java/org/apache/apex/malhar/lib/fs/s3output/S3BlockUpload.java

```diff
@@ -0,0 +1,516 @@
+  {
+    @Override
+    public void process(S3InitiateFileUpload.UploadFileMetadata tuple)
+    {
+      processUploadFileMetadata(tuple);
+    }
+  };
+
+  /**
+   * Convert each block of a given file into S3BlockMetaData
+   * @param tuple UploadFileMetadata
+   */
+  protected void processUploadFileMetadata(S3InitiateFileUpload.UploadFileMetadata tuple)
+  {
+    long[] blocks = tuple.getFileMetadata().getBlockIds();
+    String filePath = tuple.getFileMetadata().getFilePath();
+    for (int i = 0; i < blocks.length; i++) {
+      String blockId = getUniqueBlockIdFromFile(blocks[i], filePath);
+      if (blockInfo.get(blockId) != null) {
+        break;
+      }
+      blockInfo.put(blockId, new S3BlockMetaData(tuple.getKeyName(), tuple.getUploadId(), i + 1));
+    }
+    if (blocks.length > 0) {
+      blockInfo.get(getUniqueBlockIdFromFile(blocks[blocks.length - 1], filePath)).setLastBlock(true);
+    }
+    if (waitingTuples.size() > 0) {
+      processWaitBlocks();
+    }
+  }
+
+  /**
+   * Construct the unique block id from the given block id and file path.
+   * @param blockId Id of the block
+   * @param filepath given filepath
+   * @return unique block id
+   */
+  public static String getUniqueBlockIdFromFile(long blockId, String filepath)
+  {
+    return Long.toString(blockId) + "|" + filepath;
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    waitingTuples = new ArrayList<>();
+    currentWindowRecoveryState = new HashMap<>();
+    windowDataManager.setup(context);
+    s3Client = createClient();
+  }
+
+  /**
+   * Create AmazonS3 client using AWS credentials
+   * @return AmazonS3
+   */
+  protected AmazonS3 createClient()
+  {
+    AmazonS3 client = new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey));
+    if (endPoint != null) {
+      client.setEndpoint(endPoint);
+    }
+    return client;
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    currentWindowId = windowId;
+    if (windowId <= windowDataManager.getLargestCompletedWindow()) {
+      replay(windowId);
+    }
+  }
+
+  /**
+   * Replay the state.
+   * @param windowId replay window Id
+   */
+  protected void replay(long windowId)
+  {
+    try {
+      @SuppressWarnings("unchecked")
+      Map<String, UploadBlockMetadata> recoveredData = (Map<String, UploadBlockMetadata>)windowDataManager.retrieve(windowId);
+      if (recoveredData == null) {
+        return;
+      }
+      for (Map.Entry<String, UploadBlockMetadata> ubm: recoveredData.entrySet()) {
+        output.emit(ubm.getValue());
+        blockInfo.remove(ubm.getKey());
+      }
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void endWindow()
+  {
+    if (waitingTuples.size() > 0) {
+      processWaitBlocks();
+    }
+
```

Comment:
I think it will be good to check if all the blocks which were supposed to be processed in this window have indeed been processed.